### PR TITLE
Support for deploying to GitHub Releases from Travis CI

### PR DESCRIPTION
**Pull request by sjudeng**
@mbrukman This PR is just for consideration in case there's interest. It adds support for deploying JanusGraph distributions to [GitHub Releases](https://help.github.com/articles/about-releases/) as part of Travis CI builds. For example result see [this project](https://github.com/ngageoint/titan/releases). Note it only works for tags, so for example it wouldn't immediately solve #40 unless you'd want to create periodic tags (not sure if it would work to overwrite files).

If you want to try it out you'll need to follow [Travis CI instructions](https://docs.travis-ci.com/user/deployment/releases/) and update .travis.yml to change `CHANGE_ME` to the relevant encrypted api key.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -86,6 +86,21 @@ script:
   - mvn install -DskipTests=true -B -pl janusgraph-${MODULE} -am
   - mvn verify -pl janusgraph-${MODULE} ${ARGS}
 
+before_deploy:
+  - mvn package -DskipTests -Dgpg.skip=true -Pjanusgraph-release
```
Comment (amcp) [inline .travis.yml]: why skip gpg?

Comment (sjudeng) [inline .travis.yml]: I never tried to work through getting GPG signing to work as part of a Traivs CI job. Not sure if it's possible. If not maybe deploying in this way is a non-starter? I do like the general idea of handling release deployment as part of the CI process though.

Comment (amcp) [inline .travis.yml]: lets make an issue to track GPG signing in order to unblock this PR.

**Comment by amcp** [comment]
@mbrukman I am all for automatic SNAPSHOT releases. Can you take a look and if it make sense approve this?

**Comment by amcp** [comment]
@sjudeng Are these __all__ the changes necessary for GitHub Releases deployments? Do we also need to set a CHANGE_ME?

**Comment by amcp** [comment]
@mbrukman Marko Rodriguez says as soon as we make some kind of release (even if it is a snapshot) we can get listed on the Tinkerpop landing page (#107). Seems like a very good avenue for visibility so lets prioritize this action item?

**Comment by sjudeng** [comment]
@amcp Yes someone with write permissions on the repo would need to setup the encrypted API key and update .travis.yml in this PR to change `CHANGE_ME` to that key. See [Travis releases deployment instructions](https://docs.travis-ci.com/user/deployment/releases/) for more info.

But note that this only works if a tag (e.g. 0.1.0) is pushed to the repository, so it's really more for the standard release case. A snapshot tag (e.g.  or 0.1.0-SNAPSHOT-201702121349) would need to be pushed to get this to work as-is for the snapshot release case .. so it's not automatic. It looks like it is possible to create and push tags as part of the Travis build, so a fully automatic solution might be possible but isn't currently implemented in this PR.

**Comment by amcp** [comment]
@sjudeng I've got a key. Ill file a PR against this one and you can merge it into yours

**Comment by amcp** [comment]
@sjudeng now you can add the line to the build script to create the tag.

File: .travis.yml [hunk 2]
```diff
@@ -86,6 +86,21 @@ script:
...
+  - export JANUSGRAPH_DIST_ZIP="$(ls janusgraph-dist/janusgraph-dist-hadoop-2/target/janusgraph-*-hadoop2.zip)"
+  - echo "Deploying $JANUSGRAPH_DIST_ZIP to GitHub releases"
```
Comment (amcp) [inline .travis.yml]: create and push the tag here

**Comment by sjudeng** [comment]
I don't think it would go in `before_deploy` as this would only be reached if deploying (e.g. tag was already created). It could go in `after_success` (see below). But I think pushing will fail because Travis only has read access to the repo. It looks like there are examples of accomplishing write/push from Travis build, but I haven't looked into this. Until this is resolved or maybe a better alternative for snapshot releases is found, I'd recommend going with the manual tag creation at this point, either for stable release or an initial alpha "snapshot" release.

https://github.com/travis-ci/travis-ci/issues/1476
http://stackoverflow.com/questions/28217556/travis-ci-auto-tag-build-for-github-release

**Comment by amcp** [comment]
Can you configure Travis to run once a day for the purpose of creating snapshot releases?

**Comment by sjudeng** [comment]
It looks like there's support for [Cron Jobs in Travis](https://docs.travis-ci.com/user/cron-jobs). To use with the deploy capabilities I'm guessing there would need to be a conditional for `TRAVIS_EVENT_TYPE="cron"` or something like that. Regarding maybe the harder problem of enabling write/push from Travis to GitHub, it seems like using the [GitHub API](https://developer.github.com/v3/git/tags/) would be the best. One would setup a [GitHub API token](https://github.com/blog/1509-personal-api-tokens) with write access, [encrypt it with Travis](https://docs.travis-ci.com/user/encryption-keys/) and then create a script (probably parametrized by build number/etc. to be passed from Travis) to perform the tag using the API.

But again this might not be necessary for this project if standard releases, with maybe some intermediate manually-created snapshot releases, would be sufficient.

**Comment by amcp** [comment]
@sjudeng lets close out the remaining items on this PR?

**Comment by sjudeng** [comment]
Closing this for now. We're releasing infrequently and handling release artifact deployment manually at this point. Also GPG signing which isn't included here seems like it would be required for formal release artifacts.

**Pull request closed without merging**